Flexible conflict resolution

barrywhart opened the pull request:
Fixes #613 

Based on #627, but:
- Supports both old (`.rej` files) and new (inline markers) conflict behavior
- Organizes the code to reduce likelihood of merge conflicts (while this PR is in development/review)

barrywhart commented on tests/test_subdirectory.py:
```diff
@@ -152,7 +152,18 @@ def test_update_subdirectory_from_root_path(tmp_path_factory):
+        ),
+    ],
+)
+def test_new_version_uses_subdirectory(conflict, tmp_path_factory, readme):
```
Expanded this into two test cases, one for the current (`rej`) behavior and one for the new (`inline`) behavior.

barrywhart commented on tests/test_updatediff.py:
```diff
@@ -366,7 +366,7 @@ def test_commit_hooks_respected(tmp_path_factory):
         git("tag", "v2")
     # Update subproject to v2
-    copy(dst_path=dst1, defaults=True, overwrite=True)
+    copy(dst_path=dst1, defaults=True, overwrite=True, conflict="rej")
```
Updated tests in this file to explicitly specify `rej` mode. This is not strictly needed, as `rej` is the default behavior, but this will prevent the tests from breaking if the default is later changed.

barrywhart commented:
@yajo: This is ready for review. Most of the code is copied from #627; I mainly organized things a bit differently to support both old and new behavior.

As a long-time git user, I'm excited about this feature! Conflicts are scary enough, but they're less scary when presented in a familiar way.

barrywhart commented:
The failed build looks like some sort of technical GitHub or build issue, not something about the PR.

barrywhart commented:
I'm happy to remove the old behavior. Before we decide, I'll explain some things:
* In this PR, I took the code directly from #627, without studying or understanding it.
* I think this code does use 'git` to generate the conflict markers directly.

It may be possible to meet your goals of changing less code while replacing the existing behavior, but that'll be more work, possibly a *lot^ more, because it requires a deep understanding of the old and new diff algorithms, and I worry I'll get stuck, then this PR will become stale like #627 did.

My suggestions for the simplest path forward:
* Add the new diff algorithm. Make it the default behavior.
* Keep the old behavior _for now_. Deprecate it, then remove it in a future release, once the new one has been adopted successfully and proves stable. This eliminates the risk of adopting the new, not well understood diff method.

What do you think? If you like this idea, I can update the PR with the new default and the pre-commit doc change.

yajo commented:
Thanks! I understand your position and I thank you for the time you took on
updating that pr.

You can leave it as it is. I'll review when I have some time and get to the
final solution.

I prefer to maintain just one version of the algorithm though. This subject
is not really something that renders copier useless, and there's no hurry.
So don't worry, be patient and it will land well supported! ❤ 😊


El sáb., 24 sept. 2022 12:26, Barry Hart ***@***.***>
escribió:

> I'm happy to remove the old behavior. Before we decide, I'll explain some
> things:
>
>    - In this PR, I took the code directly from #627
>    <https://github.com/copier-org/copier/pull/627>, without studying or
>    understanding it.
>    - I think this code does use 'git` to generate the conflict markers
>    directly.
>
> It may be possible to meet your goals of changing less code while
> replacing the existing behavior, but that'll be more work, possibly a *lot^
> more, because it requires a deep understanding of the old and new diff
> algorithms, and I worry I'll get stuck, then this PR will become stale like
> #627 <https://github.com/copier-org/copier/pull/627> did.
>
> My suggestions for the simplest path forward:
>
>    - Add the new diff algorithm. Make it the default behavior.
>    - Keep the old behavior *for now*. Deprecate it, then remove it in a
>    future release, once the new one has been adopted successfully and proves
>    stable. This eliminates the risk of adopting the new, not well understood
>    diff method.
>
> What do you think? If you like this idea, I can update the PR with the new
> default and the pre-commit doc change.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/copier-org/copier/pull/807#issuecomment-1256946274>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHNXDJSXE7BK47TEZ567YLV73QPNANCNFSM6AAAAAAQPM6O6M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

barrywhart commented:
Ok, thanks!

barrywhart commented:
@yajo: Please let me know how things are going. At work, I may have time between other tasks to work more with this. When I suggested having two methods, I wanted to let you decide about the tradeoff (quick vs slow, 2 diff methods vs 1). Like you, I wonder if it's possible to generate the new diff format with fewer changes to the current code. Intuitively, I would think you could just run a different `git` command. (IIRC, it's `git apply` versus `git am`.)

In the meantime, this PR is much less likely to develop conflicts than the old one was, because of how I structured it. So no pressure on either of us. 😅

This is a very useful project. Thank you for helping maintain it! I'm a maintainer on another project (SQLFluff), so I know the ups and downs of that.

barrywhart commented on copier/main.py:
```diff
@@ -662,21 +695,23 @@ def run_update(self) -> None:
         ) as old_copy, TemporaryDirectory(
             prefix=f"{__name__}.recopy_diff."
         ) as new_copy:
-            old_worker = replace(
```
There are no real changes to this function. I mechanically extracted some code into separate functions and moved the `old_worker.run_copy()` up a bit. This makes the new code shorter and more similar to the old code.

barrywhart commented on copier/main.py:
```diff
@@ -745,6 +765,133 @@ def run_update(self) -> None:
             self.template.migration_tasks("after", self.subproject.template)
         )
 
+    def _do_copy(self):
```
Three new functions extracted from `Worker.apply_update()`.

barrywhart commented on copier/main.py:
```diff
@@ -745,6 +751,105 @@ def run_update(self) -> None:
             self.template.migration_tasks("after", self.subproject.template)
         )
 
+    def _apply_update_inline_conflict_markers(self):
```
This function is the only meaningful new code in the PR. It's a new/different implementation of the existing `Worker.apply_update()` function.

yajo replied:
Thanks, may you please add docstrings here too?

barrywhart replied:
Done!

barrywhart commented on copier/main.py:
```diff
@@ -745,6 +751,105 @@ def run_update(self) -> None:
+            self.template.migration_tasks("after", self.subproject.template)
+        )
+
+    def _do_copy(self):
```
These 3 functions were trivially extracted from the existing `Worker.apply_update()` function.

yajo replied:
The method name is misleading, as we already have a `run_copy()` one.

It would need some docstring and a better name. Maybe something like `_uncached_copy()`?

barrywhart replied:
Done and done!

barrywhart commented:
@yajo: I looked over the PR and found a lot of unnecessary and/or duplicate code. I deleted what I could, and I created some helper functions that reduce the duplicate code between the old and new conflict behaviors. Now there is just one function (`Worker._apply_update_inline_conflict_markers()`) that differs between the old and new behavior. I think this will make reviewing it much simpler.

I'm 💯 with removing the old behavior, but I think having them both available, at least _within this PR_ may be helpful for review and test. Once you review this and feel comfortable with the new implementation, feel free to close this and create a new PR with the old behavior removed, or ping me and I'll update this PR. (Note that a few existing tests will need updating because they currently check for the old behavior.)

barrywhart commented on copier/main.py:
```diff
@@ -745,6 +751,105 @@ def run_update(self) -> None:
+                    subfile_names.append(subfile_kind)
+
+                with local.cwd(merge_dst_temp):
+                    output = git("merge-file", "-p", *subfile_names, retcode=None)
```
```suggestion
                    # https://git-scm.com/docs/git-merge-file
                    output = git("merge-file", "-p", *subfile_names, retcode=None)
```

barrywhart commented:
@yajo: Do you think you may be able to review this soon? We'd really like to use this new behavior on our projects.

yajo commented on copier/main.py:
```diff
@@ -130,6 +145,9 @@ class Worker:
             See [quiet][].
+
+        conflict:
+            One of "rej", "inline"
```
To keep it homogeneous and warn about this new algorithm status:

```suggestion
            One of "rej" (default), "inline" (still experimental).
```

yajo commented on copier/main.py:
```diff
@@ -662,21 +681,23 @@ def run_update(self) -> None:
             )
+        self.apply_update()
+
+    def apply_update(self):
```
It seems to me like this one method should be private, as it doesn't make sense to call it directly from API, right?

If so:

```suggestion
    def _apply_update(self):
```

barrywhart replied:
Done!

yajo commented on copier/main.py:
```diff
@@ -745,6 +751,106 @@ def run_update(self) -> None:
+        # Do a normal update in final destination
+        self.run_copy()
+
+    def _make_old_worker(self, old_copy):
+        old_worker = replace(
+            self,
+            dst_path=old_copy,
+            data=self.subproject.last_answers,
+            defaults=True,
+            quiet=True,
+            src_path=self.subproject.template.url,
+            vcs_ref=self.subproject.template.commit,
+        )
+        return old_worker
+
+    def _git_initialize_repo(self):
+        git("init", retcode=None)
+        git("add", ".")
+        git("config", "user.name", "Copier")
+        git("config", "user.email", "copier@copier")
+        # 1st commit could fail if any pre-commit hook reformats code
+        git("commit", "--allow-empty", "-am", "dumb commit 1", retcode=None)
+        git("commit", "--allow-empty", "-am", "dumb commit 2")
+        git("config", "--unset", "user.name")
+        git("config", "--unset", "user.email")
```
Both methods needs docstrings, please.

yajo commented on docs/updating.md:
```diff
@@ -22,15 +22,30 @@ other Git ref you want.
+Copier to know what to do with a diff code hunk. In those cases, copier handles the
+conflict one of two ways, controlled with the `--conflict` option:
+
+-   `--conflict rej`: Creates a separate `.rej` file for each file with conflicts. These
```
```suggestion
-   `--conflict rej` (default): Creates a separate `.rej` file for each file with conflicts. These
```

yajo commented on docs/updating.md:
```diff
@@ -22,15 +22,30 @@ other Git ref you want.
+
+-   `--conflict rej`: Creates a separate `.rej` file for each file with conflicts. These
+    files contain the unresolved diffs.
+-   `--conflict inline`: Updates the file with conflict markers. This is quite similar
```
```suggestion
-   `--conflict inline` (experimental): Updates the file with conflict markers. This is quite similar
```

barrywhart replied:
Done!

yajo commented on tests/test_subdirectory.py:
```diff
@@ -209,13 +220,14 @@ def test_new_version_uses_subdirectory(tmp_path_factory):
     assert (project_path / "README.md").exists()
     with (project_path / "README.md").open() as fd:
-        assert fd.read() == "upstream version 2\n"
+        assert fd.read() == readme
```
According to what I see in this test, you should assert that a `README.md.rej` file exists when in `rej` mode. This way, we make sure that both modes detect the same conflict, and just its output mode changes.

barrywhart replied:
Done!

yajo reviewed (changes requested):
Thank you very much. The patch looks promising.

I think that we can flag the new mode as experimental and see how it goes for a few releases.

Apart from some comments below, I'm also missing docs for this new flag under [Available settings](https://copier.readthedocs.io/en/stable/configuring/#available-settings).

Sorry for taking long to review. Busy life here. Thanks for your patience!

barrywhart commented on docs/updating.md:
```diff
@@ -22,15 +22,30 @@ other Git ref you want.
+That's why the recommended way to prevent these mistakes is to add a
+[pre-commit](https://pre-commit.com/) (or equivalent) hook that forbids committing
+conflict files or markers. The recommended hook configuration depends on the `conflict`
+setting you use.
```
Done!

barrywhart commented:
Thanks for the thorough review, @yajo. I think I've addressed all of your suggestions.

I really appreciate your work maintaining this project. 🙌💪

Please let me know if you see anything else that needs to be improved.

yajo reviewed (approved):
Neat!

barrywhart commented:
Looks like a back-end issue with the code coverage reporting (likely not related to this PR):
```
Codecov report uploader 0.3.2
[2022-11-07T10:02:24.314Z] ['info'] => Project root located at: /home/runner/work/copier/copier
[2022-11-07T10:02:24.315Z] ['info'] -> No token specified or token is empty
[2022-11-07T10:02:24.323Z] ['info'] Searching for coverage files...
[2022-11-07T10:02:24.620Z] ['info'] => Found 1 possible coverage files:
  ./coverage.xml
[2022-11-07T10:02:24.621Z] ['info'] Processing ./coverage.xml...
[2022-11-07T10:02:24.625Z] ['info'] Detected GitHub Actions as the CI provider.
[2022-11-07T10:02:24.627Z] ['info'] Pinging Codecov: https://codecov.io/upload/v4?package=github-action-3.1.1-uploader-0.3.2&token=*******&branch=flexible_conflict_resolution&build=3404684625&build_url=https%3A%2F%2Fgithub.com%2Fcopier-org%2Fcopier%2Factions%2Fruns%2F3404684625&commit=f0e067871d119564053c1e78af8d7[44](https://github.com/copier-org/copier/actions/runs/3404684625/jobs/5670911211#step:12:45)58022aac1&job=CI&pr=807&service=github-actions&slug=copier-org%2Fcopier&name=copier&tag=&flags=unittests&parent=
[2022-11-07T10:02:24.919Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
```

pawamoy commented:
Codecov again 🤦 
I personally stopped using such services, they're just noise IMO.

yajo commented:
Codecov problem fixed in #861. Please rebase to get the fix.

However, the logs reveal real test failures, so please could you fix them? Thanks!

barrywhart commented:
@yajo: I rebased and pushed. I didn't check the build log before pushing, but all tests are passing locally on my Mac. Can you please trigger the build so I can see the failures? Thanks!

```
===================================================================================== 443 passed, 1 skipped, 2 xfailed, 338 warnings in 91.64s (0:01:31) =====================================================================================
```

yajo commented on tests/test_subdirectory.py:
```diff
@@ -152,7 +152,21 @@ def test_update_subdirectory_from_root_path(tmp_path_factory):
+        ("rej", "upstream version 2\n", True),
+        (
+            "inline",
+            "<<<<<<< modified\ndownstream version 1\n"
+            "=======\nupstream version 2\n>>>>>>> new upstream\n",
```
It's weird that all other `\n` work out fine. 🤔 

```suggestion
            f"<<<<<<< modified\ndownstream version 1{os.linesep}"
            "=======\nupstream version 2\n>>>>>>> new upstream\n",
```

yajo commented:
It seems the fix was not correct. Could you check out windows build error logs please.

barrywhart commented:
@yajo: I don't have a convenient Windows machine I can test on, but I added some code to normalize the line endings before comparing the strings. I'm pretty confident this will work.

yajo commented on tests/test_subdirectory.py:
```diff
@@ -209,13 +227,24 @@ def test_new_version_uses_subdirectory(tmp_path_factory):
-        assert fd.read() == "upstream version 2\n"
+
+    # On Windows, skip checking the inline conflicts due to line ending issues.
+    if conflict == "rej" or os.name != "nt":
```
Please use an appropriate xfail mark: https://github.com/copier-org/copier/blob/080e3f645e26c01ccce314634159651001dcf1cd/tests/test_migrations.py#L25-L29

barrywhart replied:
Will do. It will look a bit different because it's 1 of 2 cases failing for a parametrized test.

I followed the suggestion here?
https://stackoverflow.com/questions/30571481/how-to-mark-test-as-xfail-only-with-specific-parameters-with-pytest

yajo reviewed (changes requested):
Suggestion for better xfail marking. OTOH you have a free windows dev vm in https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/ in case you want to really fix it.

barrywhart commented:
Something strange happened with the tests. I'm going to revert my changes to the test, because the build was working before that. Can we please merge this, and then make any final adjustments to the tests in a separate PR?

barrywhart commented on tests/test_subdirectory.py:
```diff
@@ -159,7 +160,16 @@ def _normalize_line_endings(text):
     [
-        ("rej", "upstream version 2\n", True),
+        pytest.param(
+            "rej",
```
Oh wow -- silly mistake. Argh, Windows.

barrywhart replied:
Fixed!

barrywhart commented:
@yajo: I got the Windows test working with a workaround. Apparently, git is introducing line-ending issues when it's inserting inline conflict markers on Windows. I tried various settings for git's `core.autocrlf` and `.gitattributes`. Any other combination of settings resulted in one or both of:
* Duplicate newline characters (`"\n\n"`)
* Duplicate carriage-return characters (`"\r\r\n"`)

While the test does work now, perhaps we should document that caution is advised when using inline mode on Windows, or disable the feature. What do you think?

yajo commented:
The test looks ugly now, but it's better an ugly test than a skipped test. 😉 

So let's leave it as it is. The feature is flagged as experimental, so there's no need to disable it in Windows. I think that the best thing you could do with your discoveries is to open an issue for Git itself.

barrywhart commented:
I'll look into reporting the issue to git. Apparently that's done using an email list, not through GitHub like most projects.

barrywhart commented:
I reported the issue to Windows git (separate project from Unix git): https://github.com/git-for-windows/git/issues/4113
